src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClauseExtensions.cs

```diff
@@ -86,7 +86,57 @@ internal static void GetSelectExpandPaths(this SelectExpandClause selectExpandCl
         /// <returns>String list generated from selected items</returns>
         internal static List<string> GetCurrentLevelSelectList(this SelectExpandClause selectExpandClause)
         {
-            return selectExpandClause.SelectedItems.Select(GetSelectString).Where(i => i != null).ToList();
+            HashSet<string> levelSelectList = new HashSet<string>();
+            List<string> masterSelectList = new List<string>();
+
+            foreach (var selectItem in selectExpandClause.SelectedItems)
+            {
+                if (selectItem is WildcardSelectItem)
+                {
+                    levelSelectList.Add("*");
+                    continue;
+                }
+
+                NamespaceQualifiedWildcardSelectItem namespaceQualifiedWildcard = selectItem as NamespaceQualifiedWildcardSelectItem;
+                if (namespaceQualifiedWildcard != null)
+                {
+                    levelSelectList.Add(namespaceQualifiedWildcard.Namespace + ".*" );
+                    continue;
+                }
+
+                PathSelectItem pathSelectItem = selectItem as PathSelectItem;
+                if (pathSelectItem != null)
+                {
+                    IList<string> pathSelectItems = GetSelectStringFromPathSelectItem(pathSelectItem);
+
+                    for (int i = 0; i< pathSelectItems.Count; i++)
+                    {
+                        levelSelectList.Add(pathSelectItems[i]);                       
+                    }
+                }
+            }
+
+            foreach(var item in levelSelectList)
+            {

```

Comment:
For code maintainability, add a comment describing what you're doing here, including an example.  i.e., 
// If a select item is a child of a complex property, and that complex property already has all properties selected, 
// then this child is redundant and should not be included in the contextUrl.  i.e.; $select=address,address/city => address